File: webhooks/nodejs/samples/user.created.js
// HR Folder + NDA Generation + Group Assignment

const express = require("express");
const crypto = require("crypto");
const axios = require("axios");
require("dotenv").config();

const app = express();
app.use(express.json());

// Generate HMAC-SHA256 signature
function getSecretHash(secretKey, body) {
    return (
        "sha256=" +
        crypto.createHmac("sha256", secretKey).update(body).digest("hex").toUpperCase()
    );
}

app.post("/webhook", async (req, res) => {
    const sig = req.headers["x-docspace-signature-256"];
    if (!sig) {
        console.error("Unsigned request");
        return res.sendStatus(401);
    }

    const secretKey = process.env.WEBHOOK_SECRET_KEY;
    const body = JSON.stringify(req.body);
    const hash = getSecretHash(secretKey, body);

    if (sig !== hash) {
        console.error("Invalid signature");
        return res.sendStatus(401);
    }

    const event = req.body?.event?.trigger;
    if (event !== "user.created") return res.sendStatus(200);

    const user = req.body.payload;
    const fullName = `${user.firstName} ${user.lastName}`;
    const hrFolderId = process.env.HR_FOLDER_ID;

    try {
        // Step 1: Create personal HR folder
        const folderResponse = await axios.post(
            `${process.env.PORTAL_URL}/api/2.0/files/folder/${hrFolderId}`,
            {
                title: fullName,
                parentId: hrFolderId
            },
            {
                headers: {
                    Authorization: `Bearer ${process.env.API_KEY}`,
                    "Content-Type": "application/json"
                }
            }
        );

        const personalFolderId = folderResponse.data.response.id;
        console.log(`HR folder "${fullName}" created`);

        // Step 2: Copy NDA template to the user's folder
        const ndaFileTitle = `NDA_${user.firstName}_${user.lastName}.docx`;

        await axios.post(
            `${process.env.PORTAL_URL}/api/2.0/files/copy`,
            {
                fileId: process.env.NDA_TEMPLATE_FILE_ID,
                folderId: personalFolderId,
                title: ndaFileTitle
            },
            {
                headers: {
                    Authorization: `Bearer ${process.env.API_KEY}`,
                    "Content-Type": "application/json"
                }
            }
        );

        console.log(`NDA file created: ${ndaFileTitle}`);

        // Step 3: Add user to the "New Hires" group
        await axios.post(
            `${process.env.PORTAL_URL}/api/2.0/group/${process.env.NEW_HIRES_GROUP_ID}/add`,
            {
                userIds: [user.id]
            },
            {
                headers: {
                    Authorization: `Bearer ${process.env.API_KEY}`,
                    "Content-Type": "application/json"
                }
            }
        );

        console.log(`User ${fullName} added to New Hires group`);
        res.sendStatus(200);
    } catch (err) {
        console.error("Error processing user.created event:", err.response?.data || err.message);
        res.sendStatus(500);
    }
});

const PORT = process.env.PORT || 8080;
app.listen(PORT, () => {
    console.log(`Webhook server listening on port ${PORT}`);
});

File: webhooks/python/samples/user.created.py
# HR Folder + NDA Generation + Group Assignment

from flask import Flask, request, abort
import hmac
import hashlib
import requests
import os
import json

app = Flask(__name__)

# Generate HMAC-SHA256 signature
def get_secret_hash(secret_key, body):
    return "sha256=" + hmac.new(secret_key.encode(), body.encode(), hashlib.sha256).hexdigest().upper()

@app.route("/webhook", methods=["POST"])
def webhook():
    sig = request.headers.get("x-docspace-signature-256")
    if not sig:
        app.logger.error("Unsigned request")
        abort(401)

    secret_key = os.environ.get("WEBHOOK_SECRET_KEY")
    body = json.dumps(request.json)
    hash_ = get_secret_hash(secret_key, body)

    if sig != hash_:
        app.logger.error("Invalid signature")
        abort(401)

    event = request.json.get("event", {}).get("trigger")
    if event != "user.created":
        return "", 200

    user = request.json.get("payload", {})
    full_name = f"{user.get('firstName')} {user.get('lastName')}"
    hr_folder_id = os.environ.get("HR_FOLDER_ID")

    try:
        # Step 1: Create personal HR folder
        folder_response = requests.post(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/files/folder/{hr_folder_id}",
            json={
                "title": full_name,
                "parentId": hr_folder_id
            },
            headers={
                "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                "Content-Type": "application/json"
            }
        ).json()

        personal_folder_id = folder_response.get("response", {}).get("id")
        app.logger.info(f'HR folder "{full_name}" created')

        # Step 2: Copy NDA template to the user's folder
        nda_file_title = f"NDA_{user.get('firstName')}_{user.get('lastName')}.docx"

        requests.post(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/files/copy",
            json={
                "fileId": os.environ.get("NDA_TEMPLATE_FILE_ID"),
                "folderId": personal_folder_id,
                "title": nda_file_title
            },
            headers={
                "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                "Content-Type": "application/json"
            }
        )

        app.logger.info(f"NDA file created: {nda_file_title}")

        # Step 3: Add user to the "New Hires" group
        requests.post(
            f"{os.environ.get('PORTAL_URL')}/api/2.0/group/{os.environ.get('NEW_HIRES_GROUP_ID')}/add",
            json={
                "userIds": [user.get("id")]
            },
            headers={
                "Authorization": f"Bearer {os.environ.get('API_KEY')}",
                "Content-Type": "application/json"
            }
        )

        app.logger.info(f"User {full_name} added to New Hires group")

        return "", 200

    except Exception as err:
        app.logger.error(f"Error processing user.created event: {err}")
        return "", 500

if __name__ == "__main__":
    port = int(os.environ.get("PORT", 8080))
    app.run(host="0.0.0.0", port=port)
